### PR TITLE
Add support for decoding input with ffmpeg (Linux)

### PR DESCRIPTION
**Pull request by WilliamTambellini**
WIP: for early review only. Do not merge.

- add cmake option to build with ffmpeg
- search for ffmpeg at cmake time
- include ffmpegs/libav headers in main.cpp 
Remaining todos:
- if needed convert input file to wav 16khz in main.cpp

### CODE REVIEW

**Comment by WilliamTambellini** [comment]
@ggerganov @slaren  could you please early review before I move on ? Best, WT.

File: CMakeLists.txt [hunk 1]
```diff
@@ -125,6 +129,23 @@ else()
...
+  find_package(PkgConfig REQUIRED)
+  pkg_check_modules(FFMPEG REQUIRED IMPORTED_TARGET libavdevice libavfilter libavformat libavcodec libswresample libswscale libavutil)
+  if (NOT ${FFMPEG_FOUND})
+    message(FATAL_ERROR "Cannot find ffmpeg libs/headers")
+  endif()
+  message(STATUS "Found ffmpeg headers in: ${FFMPEG_INCLUDE_DIRS}")
+  include_directories(${FFMPEG_INCLUDE_DIRS})
+  add_compile_definitions(WHISPER_FFMPEG)
+  set(WHISPER_EXTRA_LIBS  ${WHISPER_EXTRA_LIBS} PkgConfig::FFMPEG)
+endif()
```
Comment (ggerganov) [inline CMakeLists.txt:147]: Use 4 spaces indent

**Comment by ggerganov** [comment]
We can add the `FindFFmpeg.cmake` script in the `cmake` folder and use it to find `ffmpeg` libs

Probably the conversion functionality should be implemented in `common.cpp` so that it can be reused by all examples, not just `main`

**Comment by WilliamTambellini** [comment]
> We can add the `FindFFmpeg.cmake` script in the `cmake` folder and use it to find `ffmpeg` libs
done

> Probably the conversion functionality should be implemented in `common.cpp` so that it can be reused by all examples, not just `main`
ok

File: CMakeLists.txt (hunk 1 above)
Comment (WilliamTambellini) [reply CMakeLists.txt:150]: done

**Comment by WilliamTambellini** [comment]
@ggerganov 
reready for review, tks

**Comment by WilliamTambellini** [comment]
@petterreinholdtsen  review please

**Comment by WilliamTambellini** [comment]
@arthw  review please

File: CMakeLists.txt [hunk 2]
```diff
@@ -59,6 +59,10 @@ option(WHISPER_BUILD_EXAMPLES         "whisper: build examples" ${WHISPER_STANDA
 
 option(WHISPER_SDL2                   "whisper: support for libSDL2" OFF)
 
+if (CMAKE_SYSTEM_NAME MATCHES "Linux")
+  option(WHISPER_FFMPEG                 "whisper: support building and linking with ffmpeg libs (avcodec, swresample, ...)" OFF)
+endif()
```
Comment (ggerganov) [inline CMakeLists.txt:64]:
```suggestion
if (CMAKE_SYSTEM_NAME MATCHES "Linux")
    option(WHISPER_FFMPEG                 "whisper: support building and linking with ffmpeg libs (avcodec, swresample, ...)" OFF)
endif()
```

File: examples/CMakeLists.txt [hunk 3]
```diff
@@ -22,13 +22,18 @@ endif()
...
+endif()
+
 add_library(${TARGET} STATIC
     common.h
     common.cpp
     common-ggml.h
     common-ggml.cpp
     grammar-parser.h
     grammar-parser.cpp
+    ${FFMPEG_SRC}
```
Comment (ggerganov) [inline examples/CMakeLists.txt:36]: For consistency, rename this variable to `COMMON_SOURCES_FFMPEG`

File: examples/common.cpp [hunk 4]
```diff
@@ -24,6 +24,11 @@
 #include <io.h>
 #endif
 
+#ifdef WHISPER_FFMPEG
+// as implemented in ffmpeg_trancode.cpp only embedded in common lib if whisper built with ffmpeg support
+extern bool ffmpeg_decode_audio(const std::string& ifname, std::vector<uint8_t>& wav_data);
```
Comment (ggerganov) [inline examples/common.cpp:29]:
```suggestion
extern bool ffmpeg_decode_audio(const std::string & ifname, std::vector<uint8_t> & wav_data);
```

File: examples/ffmpeg_transcode.cpp
Comment (ggerganov) [inline examples/ffmpeg_transcode.cpp:1]: Change the underscore in the filename to a dash for consistency: `ffmpeg-transcode.cpp`

**Review by ggerganov — APPROVED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (WilliamTambellini) [reply CMakeLists.txt:64]: done

File: examples/ffmpeg_transcode.cpp
Comment (WilliamTambellini) [reply examples/ffmpeg_transcode.cpp:1]: done

File: examples/common.cpp (hunk 4 above)
Comment (WilliamTambellini) [reply examples/common.cpp:29]: done

File: examples/CMakeLists.txt (hunk 3 above)
Comment (WilliamTambellini) [reply examples/CMakeLists.txt:36]: done

File: examples/ffmpeg_transcode.cpp
Comment (ggerganov) [reply examples/ffmpeg_transcode.cpp:1]: Hm, still uses underscore -> `ffmpeg_transcode.cpp`

**Comment by ggerganov** [comment]
Hm, I think you didn't push the correct revision - I don't see any changes since last time

**Comment by WilliamTambellini** [comment]
oops indeed. Just pushed the latest retouches.

**Comment by WilliamTambellini** [comment]
tks @ggerganov 
Any way to do a new minor release soon (eg 1.6.1) ?

**Comment by ggerganov** [comment]
done

**Comment by data-man** [comment]
Unfortunately cannot be built with FFmpeg 7.0.

**Comment by clort81** [comment]
```
/pr/Neural/Voice_Recognition_Whispr_GGML/good-whisper.cpp/examples/ffmpeg-transcode.cpp: In function ‘int decode_audio(audio_buffer*, s16**, int*)’:
/pr/Neural/Voice_Recognition_Whispr_GGML/good-whisper.cpp/examples/ffmpeg-transcode.cpp:207:5: error: ‘av_register_all’ was not declared in this scope
  207 |     av_register_all(); // from avformat. Still a must-have call for ffmpeg v3! (can be skipped for later versions)
      |     ^~~~~~~~~~~~~~~

ffmpeg                                        7:5.1.4-0+deb12u1 
```
devuan linux

**Comment by Displacer** [comment]
> ```
> 1b51fdf170714dcdd8fb9cfd02dcee684aac6150 is the first bad commit
> commit 1b51fdf170714dcdd8fb9cfd02dcee684aac6150
> Author: William Tambellini <wtambellini@sdl.com>
> Date:   Tue May 21 08:31:41 2024 -0700
> 
>     examples : add support for decoding input with ffmpeg (Linux) (#2133)
>     
> ```
> 
> ```
> /pr/Neural/Voice_Recognition_Whispr_GGML/good-whisper.cpp/examples/ffmpeg-transcode.cpp: In function ‘int decode_audio(audio_buffer*, s16**, int*)’:
> /pr/Neural/Voice_Recognition_Whispr_GGML/good-whisper.cpp/examples/ffmpeg-transcode.cpp:207:5: error: ‘av_register_all’ was not declared in this scope
>   207 |     av_register_all(); // from avformat. Still a must-have call for ffmpeg v3! (can be skipped for later versions)
>       |     ^~~~~~~~~~~~~~~
> 
> ffmpeg                                        7:5.1.4-0+deb12u1 
> ```
> 
> devuan linux

can probably be fixed with:

#if LIBAVFORMAT_VERSION_MAJOR < 56
    av_register_all(); // from avformat. Still a must-have call for ffmpeg v3! (can be skipped for later versions)
#endif

Change 56 to correct ffmpeg version. media-video/ffmpeg-4.4.4 seems to have 56 version (but i am not sure)

**Pull request merged**